### PR TITLE
feat: Allow the Helm release namespace to be overridden

### PR DESCRIPTION
**Pull request by sathieu**
for multi-namespace deployments in combined charts

### CODE REVIEW

**Comment by dnskr** [comment]
What is the use case for the changes?

What do you think about avoiding introducing `minio-operator.namespace` template and using the following code instead?
```yaml
  namespace: {{ .Values.namespaceOverride | default .Release.Namespace }}
```

**Comment by sathieu** [comment]
Thanks @dnskr for this review.

> What is the use case for the changes?

I use chart dependencies to deploy the operator and multiple tenants.

```yaml
# Chart.yaml
apiVersion: v2
name: "minio-operator"
version: 0.0.1
dependencies:
  - name: operator
    repository: https://operator.min.io/
    version: 5.0.9
  - alias: tenant1 
    name: tenant
    repository: https://operator.min.io/
    version: 5.0.9
  - alias: tenant2
    name: tenant
    repository: https://operator.min.io/
    version: 5.0.9
```

```yaml
# values.yaml
tenant1:
  namespaceOverride: minio-tenant1
tenant2:
  namespaceOverride: minio-tenant2
```


> What do you think about avoiding introducing `minio-operator.namespace` template and using the following code instead?
> 
> ```yaml
>   namespace: {{ .Values.namespaceOverride | default .Release.Namespace }}
> ```

Done.

**Comment by jiuker** [comment]
It looks like you're just using it for testing, and while that's harmless, I don't think it's necessary or best practice for helm. @pjuarezd @shtripat @harshavardhana ^^

**Comment by sathieu** [comment]
@jiuker We use this for production, not test. We are currently using [MinIO Helm chart](https://github.com/minio/minio/blob/master/helm/minio/) (i.e. not the operator), with it's limitations (including very limited pool configuration). We want to move to the operator which seems to be the recommended method.

We only have one MinIO cluster (per Kubernetes cluster), i.e. one *tenant* in the MinIO operator's terminology. This *is* good practice to use a separate namespace for the operator and each of it's tenants.

The namespaceOverride pattern is [very common](https://github.com/search?q=namespaceOverride+&type=code), as it the Helm dependencies pattern (for example [here](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack)).

We are using ArgoCD to deploy the Helm charts, and for us MinIO operator and it's tenant is only one ArgoCD application. We can move to one application for the operator and one application for each tenant, if this MR is not accepted.

**Comment by jiuker** [comment]
Yeah. Great idea. @harshavardhana @shtripat @pjuarezd cc

**Comment by shtripat** [comment]
> y one ArgoCD application.



> Yeah. Great idea. @harshavardhana @shtripat @pjuarezd cc
> 
> @sathieu And could you define as a template like promethrous do that.

@jiuker In my opinion operator is meant to manage multiple tenants. So one operator managing multiple tenants in one k8s cluster is fine. Any specific reason we want to have different operator deployments in different namespaces in same k8s cluster and they in turn maintaining their own tenants?

**Comment by jiuker** [comment]
> > y one ArgoCD application.
> 
> > Yeah. Great idea. @harshavardhana @shtripat @pjuarezd cc
> > @sathieu And could you define as a template like promethrous do that.
> 
> @jiuker In my opinion operator is meant to manage multiple tenants. So one operator managing multiple tenants in one k8s cluster is fine. Any specific reason we want to have different operator deployments in different namespaces in same k8s cluster and they in turn maintaining their own tenants?

Check the diff files. This pr only want to change the tenant. Not the operator. @shtripat

**Comment by shtripat** [comment]
> > > y one ArgoCD application.
> > 
> > 
> > > Yeah. Great idea. @harshavardhana @shtripat @pjuarezd cc
> > > @sathieu And could you define as a template like promethrous do that.
> > 
> > 
> > @jiuker In my opinion operator is meant to manage multiple tenants. So one operator managing multiple tenants in one k8s cluster is fine. Any specific reason we want to have different operator deployments in different namespaces in same k8s cluster and they in turn maintaining their own tenants?
> 
> Check the diff files. This pr only want to change the tenant. Not the operator. @shtripat

Oh ok. Then it should be fine.

**Comment by dnskr** [comment]
> We are using ArgoCD to deploy the Helm charts, and for us MinIO operator and it's tenant is only one ArgoCD application. We can move to one application for the operator and one application for each tenant, if this MR is not accepted.

@sathieu I’m sure it’s better solution, especially because you already use ArgoCD. Also consider [App Of Apps Pattern](https://argo-cd.readthedocs.io/en/stable/operator-manual/cluster-bootstrapping/#app-of-apps-pattern), which provides better management capabilities compare to chart of charts approach.

**Comment by sathieu** [comment]
> > We are using ArgoCD to deploy the Helm charts, and for us MinIO operator and it's tenant is only one ArgoCD application. We can move to one application for the operator and one application for each tenant, if this MR is not accepted.
> 
> @sathieu I’m sure it’s better solution, especially because you already use ArgoCD. Also consider [App Of Apps Pattern](https://argo-cd.readthedocs.io/en/stable/operator-manual/cluster-bootstrapping/#app-of-apps-pattern), which provides better management capabilities compare to chart of charts approach.

We'll go this path. But it can still contribute to this PR if there is value for you.

@jiuker 

> @sathieu And could you define as a template like promethrous do that.

The template was in the initial commit. See https://github.com/minio/operator/commit/5f14e772759dcb77d81331abba9d3a0a0e0e44b0. I can rollback to this if you agreee.

**Comment by jiuker** [comment]
> > > We are using ArgoCD to deploy the Helm charts, and for us MinIO operator and it's tenant is only one ArgoCD application. We can move to one application for the operator and one application for each tenant, if this MR is not accepted.
> > 
> > 
> > @sathieu I’m sure it’s better solution, especially because you already use ArgoCD. Also consider [App Of Apps Pattern](https://argo-cd.readthedocs.io/en/stable/operator-manual/cluster-bootstrapping/#app-of-apps-pattern), which provides better management capabilities compare to chart of charts approach.
> 
> We'll go this path. But it can still contribute to this PR if there is value for you.
> 
> @jiuker
> 
> > @sathieu And could you define as a template like promethrous do that.
> 
> The template was in the initial commit. See [5f14e77](https://github.com/minio/operator/commit/5f14e772759dcb77d81331abba9d3a0a0e0e44b0). I can rollback to this if you agree.

Yeah. It looks better.

File: helm/tenant/templates/NOTES.txt [hunk 1]
```diff
@@ -2,11 +2,11 @@
   To connect to the {{ .name }} tenant if it doesn't have a service exposed, you can port-forward to it by running:
   {{- if dig "certificate" "requestAutoCert" false . }}
 
-  kubectl --namespace {{ $.Release.Namespace }} port-forward svc/{{ .name }}-console 9443:9443
+  kubectl --namespace {{ $.Values.namespaceOverride | default $.Release.Namespace }} port-forward svc/{{ .name }}-console 9443:9443
```
Comment (jiuker) [inline helm/tenant/templates/NOTES.txt:5]: Define a template plz.

**Review by jiuker — APPROVED** [review]

File: helm/tenant/templates/NOTES.txt (hunk 1 above)
Comment (sathieu) [reply helm/tenant/templates/NOTES.txt]: Done.

(NB: I probably forgot to push yesterday)

**Review by jiuker — APPROVED** [review]

File: helm/tenant/templates/_helpers.tpl [hunk 2]
```diff
@@ -20,6 +20,17 @@ If release name contains chart name it will be used as a full name.
   {{- end -}}
 {{- end -}}
 
+{{/*
+Allow the release namespace to be overridden for multi-namespace deployments in combined charts
+*/}}
+{{- define "minio-operator.namespace" -}}
```
Comment (shtripat) [inline helm/tenant/templates/_helpers.tpl:26]: Now I remember, this is why I got confused earlier.
I feel we should call this variable `minio-tenant.namespace` as we are overriding the tenant namespace and not operator namespace.

Comment (sathieu) [reply helm/tenant/templates/_helpers.tpl]: Good catch @shtripat I've fixed this.

Note: the other templates in this chart are names `minio-operator.`. This should be fixed in another PR probably

**Review by jiuker — APPROVED** [review]

**Review by shtripat — APPROVED** [review]
lgtm

**Comment by dnskr** [comment]
> @dnskr We moved to another ArgoCD app. So we don't need this PR anymore.

Great! I'm using ArgoCD and App of Apps approach as well. It allows to deploy and upgrade applications in a specific order, which is not possible for chart with subcharts.

---
I found that specifying namespace in Helm charts is quite complex question without super clear answer from my point of view.
There is an issue in the helm repository with a long discussion:
- [Best Practices/Docs Question - Specify Namespace in Chart? #5465](https://github.com/helm/helm/issues/5465)

As I understand from [the first](https://github.com/helm/helm/issues/5465#issuecomment-473942223) and [the closing](https://github.com/helm/helm/issues/5465#issuecomment-692133971) answers, namespace should not be explicitly specified in Helm chart templates. All special cases like custom GitOps flows and Spinnaker behavior can be easily mitigated if they still exist.

**Comment by jiuker** [comment]
@harshavardhana Any advices for this?

**Comment by sathieu** [comment]
abandoning this change. We now install two argocd apps.

**Pull request closed without merging**